receiver/statsdreceiver/internal/protocol/statsd_parser.go

```diff
@@ -370,12 +382,75 @@ func (p *StatsDParser) Aggregate(line string, addr net.Addr) error {
 			// No action.
 		}
 	}
+}
 
-	return nil
+// statsDMetricIter allows alloc-free iteration over statsDMetric values.
+// This is helpful to support multiple values in a single statsd metric 
+// such as timers, etc.
+type statsDMetricIter struct {
+	statsDMetricProperties
+
+	// valueStr contains potentially many values separated by ":".
+	valueStr string
+	valueCount int
 }
 
-func parseMessageToMetric(line string, enableMetricType bool, enableSimpleTags bool) (statsDMetric, error) {
-	result := statsDMetric{}
+type statsDMetricProperties struct {
+	description statsDMetricDescription
+	addition    bool
+	unit        string
+	sampleRate  float64
+	timestamp   uint64
+}
+
+func (i statsDMetricIter) len() int {
+	return i.valueCount
+}
+
+func (i statsDMetricIter) metric(n int) (statsDMetric, error) {
```

Comment:
I think if you implement it as `func (i statsDMetricIter) forEach(func(v statsDMetric)) error` it'll make this trivial to do in a single pass, and still trivial to add helpers for getting all results for tests etc.

The current approach will cause us to scan the values n^2 times, as it's looping over everything twice:
```
	for i := 0; i < iter.len(); i++ {
		parseMetric, err := iter.metric(i)
/// --->
        	for j := 0; j < i.valueCount; j++ {
	     	// Cut is alloc free.
		     curr, rest, _ = strings.Cut(rest, ":")
```

... or maybe even +n as `iter.len()` will also go over the string once per `for` loop if the compiler is not smart enough.

Reply:
Yeah, can try to see if a foreach function works. Just need to be careful that it doesn’t allocate the function/lambda each time.